source/MRMesh/MRConeApproximator.h

```diff
@@ -0,0 +1,299 @@
+        cone.angle = std::acos( cosAngle );
+        cone.direction() = cone.direction().normalized();
+        cone.length = calculateConeLength( points, cone );
+
+        return;
+    }
+
+private:
+    T calculateConeLength( const std::vector<MR::Vector3<T>>& points, Cone3<T>& cone )
+    {
+        T length = static_cast< T > ( 0 );
+        for ( auto i = 0; i < points.size(); ++i )
+        {
+            length = std::max( length, std::abs( MR::dot( points[i] - cone.apex(), cone.direction() ) ) );
+        }
+        return length;
+    }
+
+    // Function for finding the best approximation of a straight line in general form y = a*x + b
+    void findBestFitLine( const std::vector<MR::Vector2<T>>& xyPairs, T& lineA, T& lineB, MR::Vector2<T>* avg = nullptr )
+    {
+        auto numPoints = xyPairs.size();
+        Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> A( numPoints, 2 );
+        Eigen::Vector<T, Eigen::Dynamic>  b( numPoints );
+
+        for ( auto i = 0; i < numPoints; ++i )
+        {
+            A( i, 0 ) = xyPairs[i].x;  // x-coordinate of the i-th point
+            A( i, 1 ) = 1.0;           // constant 1.0 for dummy term b
+            b( i ) = xyPairs[i].y;     // y-coordinate of the i-th point
+            if ( avg )
+                *avg = *avg + xyPairs[i];
+        }
+        if ( avg )
+        {
+            *avg = *avg / static_cast < T > ( xyPairs.size() );
+        }
+        // Solve the system of equations Ax = b using the least squares method
+        Eigen::Matrix<T, Eigen::Dynamic, 1> x = A.bdcSvd( Eigen::ComputeThinU | Eigen::ComputeThinV ).solve( b );
+
+        lineA = x( 0 );
+        lineB = x( 1 );
+
+        if ( avg )
+        {
+            *avg = *avg / static_cast < T > ( xyPairs.size() );
+            avg->y = lineA * avg->x + lineB;
+        }
+
+    }
+
+    void fitParamsToCone( Eigen::Vector<T, Eigen::Dynamic>& fittedParams, Cone3<T>& cone )
+    {
+        cone.apex().x = fittedParams[0];
+        cone.apex().y = fittedParams[1];
+        cone.apex().z = fittedParams[2];
+
+        cone.direction().x = fittedParams[3];
+        cone.direction().y = fittedParams[4];
+        cone.direction().z = fittedParams[5];
+    }
+    void coneToFitParams( Cone3<T>& cone, Eigen::Vector<T, Eigen::Dynamic>& fittedParams )
+    {
+        // The fittedParams guess for the cone vertex.
+        fittedParams[0] = cone.apex().x;
+        fittedParams[1] = cone.apex().y;
+        fittedParams[2] = cone.apex().z;
+
+        // The initial guess for the weighted cone axis.
+        T coneCosAngle = std::cos( cone.angle );
+        fittedParams[3] = cone.direction().x / coneCosAngle;
+        fittedParams[4] = cone.direction().y / coneCosAngle;
+        fittedParams[5] = cone.direction().z / coneCosAngle;
+    }
+
+    void ComputeInitialCone( const std::vector<MR::Vector3<T>>& points, MR::Vector3<T>& coneVertex, MR::Vector3<T>& coneAxis, T& coneAngle )
```

Comment:
should it be static returning Cone?

Reply:
will change to 

     Cone3<T> computeInitialCone_( const std::vector<MR::Vector3<T>>& points )